.cbmc-batch/source/openssl/ec_override.c

```diff
@@ -0,0 +1,443 @@
+ * zero then frees the memory associated with it. If key is NULL nothing is done.
+ */
+void EC_KEY_free(EC_KEY *key) {
+    if (key) {
+        key->references -= 1;
+        if (key->references == 0) {
+            EC_GROUP_free(key->group);
+	    BN_clear_free(key->priv_key);
+            free(key);
+        }
+    }
+}
+
+/* Could not find OpenSSL documentation. The following is taken from the header file. */
+/** Decodes a ec public key from a octet string.
+ *  \param  key  a pointer to a EC_KEY object which should be used
+ *  \param  in   memory buffer with the encoded public key
+ *  \param  len  length of the encoded public key
+ *  \return EC_KEY object with decoded public key or NULL if an error
+ *          occurred.
+ */
+EC_KEY *o2i_ECPublicKey(EC_KEY **key, const unsigned char **in, long len) {
+    assert(in);
+    assert(*in);
+    assert(AWS_MEM_IS_READABLE(*in, len));
+
+    if (!key || !(*key) || !(*key)->group || nondet_bool()) {
+        return NULL;
+    }
+
+    (*key)->pub_key_is_set = true;
+
+    // o2i_ECPublicKey calls EC_KEY_oct2key, which in some cases sets the conversion form.
+    // Can we guarantee that this never happens in our use cases?
+    // Any other possible changes to the EC_KEY?
+
+    long offset;
+    __CPROVER_assume(0 <= offset && offset <= len); // public key can be shorter than length
+    *in += offset;
+
+    return *key;
+}
+
+/* Could not find OpenSSL documentation. The following is taken from the header file. */
+/** Encodes a ec public key in an octet string.
+ *  \param  key  the EC_KEY object with the public key
+ *  \param  out  the buffer for the result (if NULL the function returns number
+ *               of bytes needed).
+ *  \return 1 on success and 0 if an error occurred
+ */
+int i2o_ECPublicKey(const EC_KEY *key, unsigned char **out) {
+  assert(ec_key_is_valid(key));
+    assert(out);           // Assuming that it's always called in ESDK with non-NULL out
+    assert(*out == NULL);  // Assuming that it's always called with NULL *out, so buffer needs to be allocated
+
+    if (!key) return 0;
+
+    int buf_len;
+    __CPROVER_assume(0 < buf_len);
+    *out = can_fail_malloc(buf_len);
+
+    if (*out == NULL) {
+        int error_code;
+        __CPROVER_assume(error_code <= 0);
+        return error_code; // Retuns 0 or negative value on error
+    }
+
+    // Should *out be incremented?
+
+    return buf_len;
+}
+
+/* Abstraction of ECDSA_SIG struct */
+struct ECDSA_SIG_st {
+  BIGNUM *r;
+  BIGNUM *s;
+};
+
+/*
+ * Description: ECDSA_SIG_get0() returns internal pointers the r and s values contained in sig and stores them in *pr and *ps, respectively. The pointer pr or ps can be NULL, in which case the corresponding value is not returned.
+ */
+void ECDSA_SIG_get0(const ECDSA_SIG *sig, const BIGNUM **pr, const BIGNUM **ps) {
+  assert(ecdsa_sig_is_valid(sig));
+  assert(pr);
+  assert(ps);
+
+  *pr = sig->r;
```

Comment:
Should we check pr/ps for null and skip the assignment as documented?

Reply:
We can do that if you think it's better, but as long as ESDK does not call it with NULL pointers the asserts should be enough.